src/backups.py

```diff
@@ -305,3 +286,214 @@ def _post_backup(self) -> Tuple[bool, str]:
+        success, error_message = self._post_restore()
+        if not success:
+            logger.warning(error_message)
+            self.charm.unit.status = BlockedStatus(error_message)
+            event.fail(error_message)
+            return
+
+        event.set_results(
+            {
+                "completed": "ok",
+            }
+        )
+
+    def _can_unit_restore_backup(self) -> Tuple[bool, str]:
+        """Validates whether this unit can restore a backup."""
+        logger.info("Checking if cluster is in blocked state")
+        if self.charm._is_cluster_blocked():
+            return False, "Cluster or unit is in a blocking state"
+
+        # TODO: Consider what happens if planned_units = 1,
+        # but there are more than 1 units in the cluster still
+        logger.info("Checking that the cluster does not have more than one unit")
+        if self.charm.app.planned_units() > 1:
+            return (
+                False,
+                "Unit cannot restore backup as there are more than one units in the cluster",
+            )
+
+        return True, None
+
+    def _pre_restore(self) -> Tuple[bool, str]:
+        """Perform operations that need to be done before performing a restore."""
+        logger.info("Setting cluster state as 'restoring'")
+        self.charm.unit_peer_data["cluster-state"] = "restoring"
+
+        logger.info(f"Stopping service {MYSQLD_SERVICE} in container {CONTAINER_NAME}")
+        # TODO: wrap around try/except for ops.model.ModelError?
+        container = self.charm.unit.get_container(CONTAINER_NAME)
+
+        try:
+            container.stop(MYSQLD_SERVICE)
+        except ChangeError as e:
+            logger.info("Setting cluster state as 'active'")
+            self.charm.unit_peer_data["cluster-state"] = "active"
+
+            error_message = f"Failed to stop service {MYSQLD_SERVICE}"
+            logger.exception(error_message, exc_info=e)
+            return False, error_message
+
+        return True, None
+
+    def _restore(self, backup_id: str, s3_parameters: Dict) -> Tuple[bool, bool, str]:
+        """Run the restore operations."""
+        datetime_restore_requested = datetime.datetime.now().strftime("%Y-%m-%dT%H:%M")
+
+        # TODO: more robustly handle raised errors
+        try:
+            logger.info("Running xbcloud get commands to retrieve the backup")
+            stdout, stderr, backup_location = self.charm._mysql.retrieve_backup_with_xbcloud(
+                s3_parameters["bucket"],
+                s3_parameters["path"],
+                s3_parameters["access-key"],
+                s3_parameters["secret-key"],
+                backup_id,
+            )
+
+            # TODO: come up with naming schema for log files
+            logfile_path = f"{s3_parameters['path'].rstrip('/')}/{datetime_restore_requested}.retrieve-{backup_id}.log"
+            self._upload_logs_to_s3(stdout, stderr, logfile_path, s3_parameters)
```

Comment:
We don't need to upload restoration logs to S3. (backups logs uploading is needed for analysis if the backup went well) 
it is enough to print restoration logs to regular charm log.

Reply:
just print them (instead of uploading restore logs to S3) in f03e92b4e45ec051635ce73a25ba64629e0eaa03